### DISCUSSION TITLE
Typesafe Argument to ProcessEvent

### DISCUSSION THREAD

**Discussion by aedenj** [nsk90/kstatemachine#71] [Q&A]
Is it possible to make the argument passed to the machine.processEvent method typesafe? E.g.

```
val machine = createStdLibStateMachine(..) {
    transition<GreenState>(...)
       .onCompleted { params, states -> (params.argument as SessionSummary).doSomething(...) } // Can I get rid of this cast?
}

// My use case is to take measurements as a result of transitions and previous states.
machine.processEventBlocking(someEvent, sessionSummary) 
```

**Comment by nsk90** [comment]
Can't you subclass Event  and put your data there as a property? Arguments API is not type safe by design.

**Reply by aedenj** [reply]
This was the conclusion I was starting to reach as a workaround, but it is that it's a work around it would appear to me. Another thought I had was to carry the measurements along with each state in which case I have the whole history. I'll probably explore the later. Thank you.
